ml/train/losses/losses.go

```diff
@@ -398,3 +398,211 @@ func MakeHuberLoss(delta float64) LossFn {
 		return loss
 	}
 }
+
+// Computes the pairwise distance matrix with numerical stability.
+
+//     output[i, j] = || embeddings[i, :] - embeddings[j, :] ||_2
+
+//     Args:
+//       embeddings: 2-D Tensor of size `[number of data, embeddings dimension]`.
+//       squared: Boolean, whether or not to square the pairwise distances.
+
+// Returns:
+//
+//	pairwise distance: 2-D Tensor of size `[number of data, number of data]`.
+func pairwiseDistance(embeddings *Node, squared bool) *Node {
+	// Distances ||a - b||^2 = ||a||^2  - 2 <a, b> + ||b||^2
+	pairwiseDistancesSquared := Add(
+		Add(
+			L2NormSquare(embeddings, 1),
+			L2NormSquare(Transpose(embeddings, 0, 1), 0),
+		),
+		MulScalar(MatMul(embeddings, Transpose(embeddings, 0, 1)), -2.0),
+	)
+	// Deal with numerical inaccuracies. Set small negatives to zero.
+	pairwiseDistancesSquaredZeros := Max(pairwiseDistancesSquared, ZerosLike(pairwiseDistancesSquared))
+	// Optionally take the sqrt.
+	pairwiseDistances := pairwiseDistancesSquaredZeros
+	if !squared {
+		// Get the mask where the zero distances are at.
+		zeros := ZerosLike(pairwiseDistancesSquaredZeros)
```

Comment:
I suggest using a scalar zero here, just in case. It may save some accelerator memory for  not creating another tensor -- although I think XLA will optimize it away.

```
g := embeddings.Graph()
dtype := embeddings.DType()
zero := ScalarZero(g, dtype)
```

You can also use `g` and `dtype` below. I usually define `g` and `dtype` at the very start of the functions where I need them.